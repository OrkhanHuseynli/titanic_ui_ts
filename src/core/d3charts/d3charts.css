#app {
    padding: 20px;
    font-family: robot, arial
}
svg {
    /*fill: #000;*/
    fill-opacity: 1;
}
path {
    fill: none;
}
path.curve-line {
    stroke: #2196f3;
    stroke-width: 2px;
    stroke-opacity: 0.8;
}

path.border-line {
    /*stroke: #03dac6;*/
    stroke: #bbdefb;
    stroke-width: 2px;
    fill: none;
}

.axis {
    stroke: #000;
}

.axis-labels text {
    fill: #000;
    fill-opacity: 0.9;
    font-size: 12px;
    text-anchor: middle;
}

.axis-labels line {
    stroke: #000;
}

.gridline {
    opacity: 0.2;
}

.circle {
    fill : #2196f3;
    stroke: #1565c0;
}

.circle:hover{
    fill : #0d4d71;
    stroke: #0d4d71;
}

.tooltip-blue {
    fill: #1565c0;
}

.tooltip-red {
    fill: #2196f3;
    /*background-color: #f44336;*/
}

.tooltip-thresh{
    opacity: 1;
    fill-opacity: 1;
    fill: #0d47a1;
    /*stroke-width: 1;*/
    /*stroke: #757e91;*/
}

.tooltip-text {
    font-family: Verdana;
    font-size: 12px;
    fill-opacity: 1;
    fill: white;
}