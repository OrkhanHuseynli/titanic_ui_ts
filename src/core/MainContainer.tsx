import React from 'react';
import {Component} from "react";
import {Button, createStyles, Grid, makeStyles, Theme, Typography} from "@material-ui/core";
import {DropZone} from "./file_upload/DropZone";
import DataTable from "./table/DataTable";
import {LineChart} from "./d3charts/d3charts";
import {isBoolean} from "util";


const width = 500, height = 350, margin = 20;
const mockData = [
    {a: -0.59, b: 0.49},
    {a: 0.25, b: 0.33},
    {a: 0.46, b: 0.27},
    {a: 0.64, b: 0.20},
    {a: 0.73, b: 0.17},
    {a: 0.89, b: 0.08},
    {a: 1.0, b: 0.0},
    {a: 1.0, b: 0.0}

];

type MainContainerState = {
    uploadStatus: boolean;
    uploadInfo: UploadInfo;
}
type UploadInfo = {
    fileName: string
    dataSize: string
    inputParamsCount: string
    outputParamsCount: string
}
function createData(name: string, value: string) {
    return { name, value };
}

const mockRows = [
    createData('File name', "titanic.csv"),
    createData('Data size', "45"),
    createData('# of input params', "3"),
    createData('# of output params', "1"),
];
export default class MainContainer extends Component<{classes: any}, MainContainerState> {
    constructor(props: any) {
        super(props);
        this.state = {
            uploadStatus: false,
            uploadInfo: {fileName: "", dataSize: "", inputParamsCount: "", outputParamsCount: ""}
        };
        this.getUploadStatus = this.getUploadStatus.bind(this);
    }


    getUploadStatus = (status: boolean):void => {
        console.debug("Upload status: " + this.state.uploadStatus)
        this.setState({uploadStatus: status})
    };

    displayDataPreviewSection = (): JSX.Element => {
            if (this.state.uploadStatus){
                return <DataTable dataRows={mockRows} />
            }
        return <div className={this.props.classes.dataPreview}><Typography color={"textSecondary"}>No data provided</Typography></div>
        };


    render() {
        return (
            <div className={this.props.classes.root}>
                <Grid container spacing={3}>
                    <Grid item xs={12}>
                        <div className={this.props.classes.paper}>
                            <Typography variant="h5"> 1. Upload data </Typography>
                        </div>
                    </Grid>
                    <Grid item xs={12}>
                        <div className={this.props.classes.paper}><DropZone getUploadStatus={this.getUploadStatus}/></div>
                    </Grid>
                    <Grid item xs={12}>
                        <div className={this.props.classes.paper}>
                            <Typography variant="h5"> 2. Data preview</Typography>
                        </div>
                    </Grid>
                    <Grid item xs={12} container  direction="row" justify="center" alignItems="center">
                        <Grid item xs={4}>
                            {this.displayDataPreviewSection()}
                        </Grid>
                    </Grid>
                    <Grid item xs={12}>
                        <div className={this.props.classes.paper}>
                            <Typography variant="h5"> 3. Train model</Typography>
                        </div>
                    </Grid>
                    <Grid item xs={12} spacing={3} container
                          direction="row"
                          justify="center"
                          alignItems="center">
                        <Button variant="contained" color="primary" className={this.props.classes.button} disabled={!this.state.uploadStatus} >
                            Start
                        </Button>
                    </Grid>
                    <Grid item xs={12} spacing={3} container
                          direction="row"
                          justify="center"
                          alignItems="center">
                        <div className={this.props.classes.paper}>
                            <Typography variant="h5"> 4. ROC curve </Typography>
                        </div>
                    </Grid>
                    <Grid item xs={12}>
                        <div className={this.props.classes.paper}>
                            <LineChart data={mockData} width={width} height={height}/>
                        </div>
                    </Grid>
                </Grid>
            </div>);
    }
}