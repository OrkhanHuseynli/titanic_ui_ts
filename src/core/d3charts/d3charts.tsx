import * as React from "react";
import * as d3 from 'd3'
import {Component} from "react";
import './d3charts.css';
import {ConfusionMatrix, ConfusionMatrixObjectList, RocPoint, RocPointList} from "../MainContainer";

type CurveProps = {
    width: any
    height: any
    rocPointList: RocPointList
    confMatrixObjList: ConfusionMatrixObjectList
    thresholdList: number[]
}
type CurveState = {
    tooltip: Tooltip
}

type Tooltip = {
    x: number
    y: number
    display: string
    width: number
    threshold: number
    confusionMatrix: ConfusionMatrix
}

type CircleData = { cx: number, cy: number, threshold: number, confusionMatrix: ConfusionMatrix }

const getCircleData = (rocPoint: RocPoint, threshold: number, confusionMatrix: ConfusionMatrix): CircleData => {
    return {cx: rocPoint.b, cy: rocPoint.a, threshold:threshold, confusionMatrix: confusionMatrix}
};

const getCircleDataList = (rocPointList: RocPointList, thresholdList: number[], confusionMatrixObjList: ConfusionMatrixObjectList): Array<CircleData> => {
    console.info("getCircleDataList");
    if (rocPointList.length !== confusionMatrixObjList.length) {
        let msg = "Length of RocPointList and ConfusionMatrixObjectList objects must be equal!";
        console.error(msg);
        throw Error(msg);
    }

    let circleDataList: Array<CircleData> = [];
    for (let i = 0; i < rocPointList.length; i++) {
        circleDataList.push(getCircleData(rocPointList[i], thresholdList[i], confusionMatrixObjList[i]))
    }
    return circleDataList;
};

export class LineChart extends Component<CurveProps, CurveState> {

    constructor(props: CurveProps) {
        super(props);
        this.state = {
            tooltip: {x: 0, y: 0, display: "none", width: 35, threshold: 0, confusionMatrix: {TP: 0, TN: 0, FP: 0, FN: 0}}
        }
    }

    onMouseOverEv = (e: any, x: number, y: number, threshold: number, confusionMatrix: ConfusionMatrix) => {
        let tooltip: Tooltip = {
            x: x,
            y: y,
            display: "true",
            width: 35,
            threshold: threshold,
            confusionMatrix: {TP: confusionMatrix.TP, TN: confusionMatrix.TN, FP: confusionMatrix.FP, FN: confusionMatrix.FN}
        };
        this.setState({tooltip: {...tooltip}})
    };

    onMouseOutEv = (e: any) => {
        let tooltip: Tooltip = {x: 0, y: 0, display: "none", width: 35, threshold: 0, confusionMatrix: {TP: 0, TN: 0, FP: 0, FN: 0}};
        this.setState({tooltip: {...tooltip}})
    };

    render() {

        const margin = 20;
        const {width, height, rocPointList, confMatrixObjList, thresholdList} = this.props;

        const h = height - 2 * margin, w = width - 2 * margin;

        //number formatter
        const xFormat = d3.format('.2');

        //x scale
        const x = d3.scaleLinear()
        // @ts-ignore
            .domain(d3.extent(rocPointList, d => d.b)) //domain: [min,max] of a
            .range([margin, w]);

        //y scale

        const y = d3.scaleLinear()
        // @ts-ignore
            .domain([0, d3.max(rocPointList, d => d.a)]) // domain [0,max] of b (start from 0)
            .range([h, margin]);

        //line generator: each point is [x(d.a), y(d.b)] where d is a row in data
        // and x, y are scales (e.g. x(10) returns pixel value of 10 scaled by x)
        const line = d3.line()
        // @ts-ignore
            .x(d => x(d.b))
            // @ts-ignore
            .y(d => y(d.a))
            .curve(d3.curveCatmullRom.alpha(0.5)); //curve line

        // @ts-ignore
        const xTicks = x.ticks(6).map(d => (
            x(d) > margin && x(d) < w ?
                <g transform={`translate(${x(d)},${h + margin})`}>
                    <text>{xFormat(d)}</text>
                    <line x1='0' y1='0' y2='5' transform="translate(0,-20)"/>
                </g>
                : null
        ));


        const yTicks = y.ticks(5).map(
            // @ts-ignore
            d => (
                y(d) > 10 && y(d) < h ?
                    <g transform={`translate(${margin},${y(d)})`}>
                        <text x="-12" y="5">{xFormat(d)}</text>
                        <line x1='0' y1='0' y2='0' transform="translate(-5,0)"/>
                        <line className='gridline' x1={w - margin} y1='0' y2='0' transform="translate(-5,0)"/>
                    </g>
                    : null
            ));

        let circlesData: Array<CircleData> = getCircleDataList(rocPointList, thresholdList, confMatrixObjList);

        let circlesSVG = <g>
            {
                circlesData.map((d: CircleData) => {
                    // @ts-ignore
                    return <g>
                        <circle className={"circle"} key={d.cx} cx={x(d.cx)} cy={y(d.cy)} r={6} onMouseOver={(e: any) => {
                            this.onMouseOverEv(e, d.cx, d.cy, d.threshold, d.confusionMatrix)
                        }} onMouseOut={this.onMouseOutEv}/>
                    </g>
                })
            }
        </g>;

        let tulAxisDistX = 0.08;
        let tulTextDistX = 0.02;
        let tulAxisDistY = 0.085;
        let tulTextDistY = 0.05;
        let thresholdDistY = 0.07;


        let tooltipSVG = <g>
            <rect className={"tooltip-thresh"} x={x(this.state.tooltip.x)} y={y(this.state.tooltip.y + thresholdDistY)}
                  display={this.state.tooltip.display} width={this.state.tooltip.width*2} height="25"/>
            <text className={"tooltip-text"} x={x(this.state.tooltip.x + tulTextDistX-0.01)}
                  y={y(this.state.tooltip.y + thresholdDistY - tulTextDistY)}
                  display={this.state.tooltip.display}>thrld : {this.state.tooltip.threshold}
            </text>
            <rect className={"tooltip-blue"} x={x(this.state.tooltip.x)} y={y(this.state.tooltip.y)}
                  display={this.state.tooltip.display} width={this.state.tooltip.width} height="25"/>
            <text className={"tooltip-text"} x={x(this.state.tooltip.x + tulTextDistX)}
                  y={y(this.state.tooltip.y - tulTextDistY)}
                  display={this.state.tooltip.display}>{this.state.tooltip.confusionMatrix.TP}
            </text>
            <rect className={"tooltip-red"} x={x(this.state.tooltip.x + tulAxisDistX)} y={y(this.state.tooltip.y)}
                  display={this.state.tooltip.display} width={this.state.tooltip.width} height="25"/>
            <text className={"tooltip-text"} fontFamily={"Verdana"} fontSize={"12"} fill={"black"}
                  x={x(this.state.tooltip.x + tulAxisDistX + tulTextDistX)}
                  y={y(this.state.tooltip.y - tulTextDistY)}
                  display={this.state.tooltip.display}>{this.state.tooltip.confusionMatrix.TN}
            </text>
            <rect className={"tooltip-red"} x={x(this.state.tooltip.x)} y={y(this.state.tooltip.y - tulAxisDistY)}
                  display={this.state.tooltip.display} width={this.state.tooltip.width} height="25"/>
            <text  className={"tooltip-text"} fontFamily={"Verdana"} fontSize={"12"} fill={"black"} x={x(this.state.tooltip.x + tulTextDistX)}
                  y={y(this.state.tooltip.y - tulAxisDistY - tulTextDistY)}
                  display={this.state.tooltip.display}>{this.state.tooltip.confusionMatrix.FN}
            </text>
            <rect className={"tooltip-blue"} x={x(this.state.tooltip.x + tulAxisDistX)}
                  y={y(this.state.tooltip.y - tulAxisDistY)}
                  display={this.state.tooltip.display} width={this.state.tooltip.width} height="25"/>
            <text className={"tooltip-text"} fontFamily={"Verdana"} fontSize={"12"} fill={"black"}
                  x={x(this.state.tooltip.x + tulAxisDistX + tulTextDistX)}
                  y={y(this.state.tooltip.y - tulAxisDistY - tulTextDistY)}
                  display={this.state.tooltip.display}>{this.state.tooltip.confusionMatrix.FP}
            </text>
        </g>

        return (
            <div>
                <svg width={width} height={height}>
                    <line className="axis" x1={margin} x2={w} y1={h} y2={h}/>
                    <line className="axis" x1={margin} x2={margin} y1={margin} y2={h}/>
                    // @ts-ignore
                    <path className={"curve-line"} d={line(rocPointList)}/>
                    // @ts-ignore
                    <path className={"border-line"} d={line([{a: 0, b: 0}, {a: 1, b: 1}])}/>
                    <g className="axis-labels">
                        {xTicks}
                    </g>
                    <g className="axis-labels">
                        {yTicks}
                    </g>
                    {circlesSVG}
                    {tooltipSVG}
                </svg>
            </div>
        )
    }
}