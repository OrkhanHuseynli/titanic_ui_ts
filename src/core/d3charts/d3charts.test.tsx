import {configure, shallow, ShallowWrapper} from "enzyme";
import { create, ReactTestRenderer } from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16'

import * as React from "react";
import {CurveProps, LineChart} from "./d3charts";
import {ConfusionMatrix, RocPointList} from "../MainContainer";

let wrapper: ShallowWrapper<CurveProps>;
let snapshot: ReactTestRenderer;

beforeEach(() => {
    const mockRocPointList: RocPointList = [{a: 0.3, b:0.5}];
    const mockConfMatrixObjList: Array<ConfusionMatrix> = [{FN: 2, FP:3, TN:0, TP:8}];
    const mockThresholdList = [0.2];
    const width = 30, height = 30;
    const mockLineChart = <LineChart rocPointList={mockRocPointList} confMatrixObjList={mockConfMatrixObjList} thresholdList={mockThresholdList} width={width} height={height}/>;

    wrapper = shallow(mockLineChart);
    snapshot = create(mockLineChart);
});

configure({adapter: new Adapter()});

describe('d3charts', () => {
    test('it matches the snapshot', () => {
        expect(snapshot.toJSON()).toMatchSnapshot();
    });
});