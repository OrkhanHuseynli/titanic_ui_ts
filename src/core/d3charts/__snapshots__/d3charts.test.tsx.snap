// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`d3charts it matches the snapshot 1`] = `
<div>
  <svg
    height={30}
    width={30}
  >
    <line
      className="axis"
      x1={20}
      x2={-10}
      y1={-10}
      y2={-10}
    />
    <line
      className="axis"
      x1={20}
      x2={20}
      y1={20}
      y2={-10}
    />
    // @ts-ignore
    <path
      className="curve-line"
      d="M5,20Z"
    />
    // @ts-ignore
    <path
      className="border-line"
      d="M5,-10L5,90"
    />
    <g
      className="axis-labels"
    />
    <g
      className="axis-labels"
    />
    <g>
      <g>
        <circle
          className="circle"
          cx={5}
          cy={20}
          onMouseOut={[Function]}
          onMouseOver={[Function]}
          r={6}
        />
      </g>
    </g>
    <g>
      <rect
        className="tooltip-thresh"
        display="none"
        height="25"
        width={70}
        x={5}
        y={-2.999999999999999}
      />
      <text
        className="tooltip-text"
        display="none"
        x={5}
        y={-8}
      >
        thrld : 
        0
      </text>
      <rect
        className="tooltip-blue"
        display="none"
        height="25"
        width={35}
        x={5}
        y={-10}
      />
      <text
        className="tooltip-text"
        display="none"
        x={5}
        y={-15}
      >
        0
      </text>
      <rect
        className="tooltip-red"
        display="none"
        height="25"
        width={35}
        x={5}
        y={-10}
      />
      <text
        className="tooltip-text"
        display="none"
        fill="black"
        fontFamily="Verdana"
        fontSize="12"
        x={5}
        y={-15}
      >
        0
      </text>
      <rect
        className="tooltip-red"
        display="none"
        height="25"
        width={35}
        x={5}
        y={-18.5}
      />
      <text
        className="tooltip-text"
        display="none"
        fill="black"
        fontFamily="Verdana"
        fontSize="12"
        x={5}
        y={-23.5}
      >
        0
      </text>
      <rect
        className="tooltip-blue"
        display="none"
        height="25"
        width={35}
        x={5}
        y={-18.5}
      />
      <text
        className="tooltip-text"
        display="none"
        fill="black"
        fontFamily="Verdana"
        fontSize="12"
        x={5}
        y={-23.5}
      >
        0
      </text>
    </g>
  </svg>
</div>
`;
